File: minigit-5/ts/src/commands/add.ts
// add command - Stage files for commit

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePath, getFileMode } from '../utils';
import { createBlob } from '../objects';
import { readIndex, writeIndex, getAllIndexEntries } from '../index-file';
import { getHeadCommit, resolveRevision } from '../refs';
import { walkTree, getTreeFromTreeIsh } from '../objects';
import { IndexEntry } from '../types';

export function add(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse flags
  let updateOnly = false;
  let addAll = false;
  const paths: string[] = [];

  for (const arg of args) {
    if (arg === '-u' || arg === '--update') {
      updateOnly = true;
    } else if (arg === '-A' || arg === '--all') {
      addAll = true;
    } else {
      paths.push(arg);
    }
  }

  // If no paths and no flags, error
  if (paths.length === 0 && !addAll && !updateOnly) {
    console.error('Nothing specified, nothing added.');
    return 1;
  }

  // Get current tracked files from HEAD commit
  const trackedFiles = new Set<string>();
  const headSha = getHeadCommit(repoRoot);
  if (headSha) {
    try {
      const treeSha = getTreeFromTreeIsh(headSha, repoRoot);
      const treeFiles = walkTree(treeSha, '', repoRoot);
      treeFiles.forEach((_, filePath) => trackedFiles.add(filePath));
    } catch {
      // No tree yet
    }
  }

  // Also add files that are in the index
  const currentIndex = readIndex(repoRoot);
  for (const entry of currentIndex.entries) {
    trackedFiles.add(entry.path);
  }

  // Collect files to add
  const filesToAdd: string[] = [];
  const filesToRemove: string[] = [];

  if (addAll || updateOnly) {
    // -A: stage all changes (new, modified, deleted)
    // -u: stage only tracked files (modified, deleted)

    // Find all files in working tree
    const workingTreeFiles = collectAllFiles(repoRoot, repoRoot);

    if (addAll) {
      // Add all working tree files
      filesToAdd.push(...workingTreeFiles);

      // Mark deleted tracked files for removal
      for (const trackedFile of trackedFiles) {
        const fullPath = path.join(repoRoot, trackedFile);
        if (!fs.existsSync(fullPath)) {
          filesToRemove.push(trackedFile);
        }
      }
    } else {
      // Update only - only tracked files
      for (const file of workingTreeFiles) {
        if (trackedFiles.has(file)) {
          filesToAdd.push(file);
        }
      }

      // Mark deleted tracked files for removal
      for (const trackedFile of trackedFiles) {
        const fullPath = path.join(repoRoot, trackedFile);
        if (!fs.existsSync(fullPath)) {
          filesToRemove.push(trackedFile);
        }
      }
    }
  } else {
    // Add specific paths
    for (const p of paths) {
      const resolvedPath = path.resolve(process.cwd(), p);

      if (!fs.existsSync(resolvedPath)) {
        // Check if it's a tracked file that was deleted
        const relativePath = normalizePath(path.relative(repoRoot, resolvedPath));
        if (trackedFiles.has(relativePath)) {
          filesToRemove.push(relativePath);
          continue;
        }
        console.error(`fatal: pathspec '${p}' did not match any files`);
        return 1;
      }

      const stats = fs.statSync(resolvedPath);
      if (stats.isDirectory()) {
        // Add all files in directory
        const dirFiles = collectAllFiles(resolvedPath, repoRoot);
        filesToAdd.push(...dirFiles);
      } else {
        const relativePath = normalizePath(path.relative(repoRoot, resolvedPath));
        filesToAdd.push(relativePath);
      }
    }
  }

  // Read current index
  const index = readIndex(repoRoot);

  // Remove deleted files from index
  for (const filePath of filesToRemove) {
    const idx = index.entries.findIndex(e => e.path === filePath);
    if (idx >= 0) {
      index.entries.splice(idx, 1);
    }
  }

  // Add/update files in index
  for (const filePath of filesToAdd) {
    const fullPath = path.join(repoRoot, filePath);

    if (!fs.existsSync(fullPath)) {
      continue;
    }

    const stats = fs.lstatSync(fullPath);

    // Skip directories
    if (stats.isDirectory()) {
      continue;
    }

    // Read file content
    let content: Buffer;
    let mode: number;

    if (stats.isSymbolicLink()) {
      // For symlinks, store the link target as content
      const target = fs.readlinkSync(fullPath);
      content = Buffer.from(target);
      mode = 0o120000;
    } else {
      content = fs.readFileSync(fullPath);
      mode = getFileMode(fullPath);
    }

    // Create blob object
    const sha = createBlob(content, true, repoRoot);

    // Update or add index entry
    const existingIdx = index.entries.findIndex(e => e.path === filePath);
    const entry: IndexEntry = {
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode,
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha,
      flags: Math.min(filePath.length, 0xfff),
      path: filePath,
    };

    if (existingIdx >= 0) {
      index.entries[existingIdx] = entry;
    } else {
      index.entries.push(entry);
    }
  }

  // Write updated index
  writeIndex(index, repoRoot);

  return 0;
}

function collectAllFiles(dir: string, repoRoot: string): string[] {
  const result: string[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    // Skip .minigit directory
    if (entry.name === '.minigit') continue;

    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      result.push(...collectAllFiles(fullPath, repoRoot));
    } else {
      const relativePath = normalizePath(path.relative(repoRoot, fullPath));
      result.push(relativePath);
    }
  }

  return result;
}

File: minigit-5/py/commands/add.py
"""add command - Stage files for commit."""

import sys
import os
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_file_mode
from objects import Blob
from index import Index, create_entry_from_file


def run(args: list[str]) -> int:
    """Stage files for commit."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    if not args:
        print("fatal: nothing specified, nothing added", file=sys.stderr)
        return 1

    # Parse flags
    all_flag = False
    update_flag = False
    paths = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ('-A', '--all'):
            all_flag = True
        elif arg in ('-u', '--update'):
            update_flag = True
        elif arg == '--':
            paths.extend(args[i + 1:])
            break
        else:
            paths.append(arg)
        i += 1

    index = Index.read(repo_root)

    if all_flag or update_flag:
        # Handle -A and -u flags
        return handle_auto_add(repo_root, index, all_flag, update_flag)

    # Add specified paths
    for path_str in paths:
        path = Path(path_str)
        if not path.is_absolute():
            path = Path.cwd() / path

        if not path.exists():
            print(f"fatal: pathspec '{path_str}' did not match any files", file=sys.stderr)
            return 1

        if path.is_file():
            add_file(repo_root, index, path)
        elif path.is_dir():
            add_directory(repo_root, index, path)

    index.write(repo_root)
    return 0


def add_file(repo_root: Path, index: Index, file_path: Path):
    """Add a single file to the index."""
    try:
        rel_path = file_path.resolve().relative_to(repo_root.resolve())
    except ValueError:
        return

    rel_path_str = str(rel_path).replace(os.sep, '/')

    # Skip .minigit directory
    if '.minigit' in rel_path.parts:
        return

    # Read file and create blob
    if file_path.is_symlink():
        # For symlinks, store the target path
        content = os.readlink(file_path).encode()
        mode = 0o120000
    else:
        content = file_path.read_bytes()
        mode = get_file_mode(file_path)

    blob = Blob(content)
    blob.write(repo_root)

    # Create index entry
    entry = create_entry_from_file(file_path, rel_path_str, blob.sha, mode)
    index.add_entry(entry)


def add_directory(repo_root: Path, index: Index, dir_path: Path):
    """Recursively add a directory to the index."""
    for item in dir_path.rglob('*'):
        if item.is_file() and '.minigit' not in item.parts:
            add_file(repo_root, index, item)


def handle_auto_add(repo_root: Path, index: Index, all_flag: bool, update_flag: bool) -> int:
    """Handle -A and -u flags."""
    # Get currently tracked files
    tracked_files = set(index.entries.keys())

    # Get all files in working directory
    work_files = set()
    for item in repo_root.rglob('*'):
        if item.is_file() and '.minigit' not in item.parts:
            try:
                rel_path = item.relative_to(repo_root)
                work_files.add(str(rel_path).replace(os.sep, '/'))
            except ValueError:
                pass

    if all_flag:
        # Stage all changes: new, modified, deleted
        # Add/update existing files
        for rel_path_str in work_files:
            file_path = repo_root / rel_path_str
            add_file(repo_root, index, file_path)

        # Remove deleted files from index
        for tracked in tracked_files:
            if tracked not in work_files:
                index.remove_entry(tracked)

    elif update_flag:
        # Only update tracked files (modified and deleted, not new)
        for tracked in tracked_files:
            file_path = repo_root / tracked
            if file_path.exists():
                add_file(repo_root, index, file_path)
            else:
                index.remove_entry(tracked)

    index.write(repo_root)
    return 0
